Objects/odictobject.c

```diff
@@ -806,30 +807,50 @@ _odict_keys_equal(PyODictObject *a, PyODictObject *b)
 {
     _ODictNode *node_a, *node_b;
 
+    // keep operands' state and size to detect undesired mutations
+    const Py_ssize_t size_a = PyODict_SIZE(a);
+    const Py_ssize_t size_b = PyODict_SIZE(b);
+    if (size_a == -1 || size_b == -1) {
+        return -1;
+    }
+    if (size_a != size_b) {
```

Comment:
Should not `dict.__eq__` already check this?

Reply:
Yes, but it does not mean that in-between `dict.__eq__` did not cause side-effects (just use a value that mutates its container and it would still pass). The previous version did:

```C
        else if (node_a == NULL || node_b == NULL)
            /* unequal length */
			return 0;
```  

So it was not necessarily the case that the length is equal. However, I will revert this line because it's maybe easier to just leave everything to the while loop instead.